### DISCUSSION TITLE
Getting results back from Shortcuts

### DISCUSSION THREAD

**Discussion by Younghart** [ttscoff/bunch-docs#241] [Ideas]
Hello everyone,

is there a way to run a shortcut and get the result back, e.g. to define a variable?

What I'd wish for would be something like

    projectname = /usr/bin/shortcuts run "Project Setup"

or something similar involving one or more less-than, greater-than, or pipe symbols. 
My workaround right now is to add a "copy to clipboard" action to my shortcut and finagle it into the Bunch process the following way:

    /usr/bin/shortcuts run "Project Setup"
    projectname = ?"What did you call your project again? (Hint: Cmd+V)"

Has anybody found a more elegant solution?

**Comment by RosemaryOrchard** [comment]
Make sure your Shortcut is returning data. On iOS by default, it always returns whatever is at the end of your Shortcut. However, on the Mac you need to use the "Stop and Output" action to return data.

**Comment by ttscoff** [comment]
First, I guess I didn't even realize the `shortcut` tool could get a response. If that's the case, I'll build that in to Bunch.

BTW, you can run a shortcut with `(shortcut project setup)` instead of calling a shell command, but if you're able to get data back that way then the shell cut makes more sense until I build the response in.

Also, you can get the contents of the clipboard into a variable with:

    projectname = $ pbpaste

Which would smooth out your workaround in case you can't actually get data back through the command line utility.

**Reply by Younghart** [reply]
That's a great shortcut within the workaround ;-) I did that right away, and it works like a charm.
I initially had problems launching shortcuts from within parentheses, so I switched to the CLI version. Right now, none of the options allow me to launch a Shortcut from within a Bunch, but I'm sure a reboot and the removal of the stupid typo I'll discover right after clicking on "Reply" will get that out of the way.
Thanks again!

**Reply by ttscoff** [reply]
I'm curious what issues you ran into using the parenthetical command? Just no response at all? Anything in the log?

**Reply by Younghart** [reply]
Sure thing! This is the original Bunch code:
    ---
    title: 📜 Neuer Artikel
    toggles: false
    ---
    
   (shortcut Artikelneuanlage)
   projectname = $ pbpaste
   Artikelname = ?"Wie hieß der Artikel nochmal? (Tipp: ${projectname})"

Result: Shortcut doesn't start, and pasteboard won't get updated (of course).
Here's the console log:

    2022-08-07 18:33:30: 🟡Debug Level set to 3. Restart required to enable.
    2022-08-07 18:33:34: ⚪Launched ==========================================
    2022-08-07 18:35:00: ⚪[Neuanlage_Artikel] Open >>>>>>>>>>>>>>>>>>>>>
    2022-08-07 18:35:26: ⚪[Neuanlage_Artikel] Processing command "shortcut Artikelneuanlage"

Also, it looks like the interactive dialogue is called before the (shortcut [])

**Reply by Younghart** [reply]
Another test ...
macOS 12.5, localization DE_de
(Bunch 1.4.9 (148)) - "Launch items in parallel" is deactivated.
The freshly created shortcut creates a timestamp, copies it to the clipboard, and Stops&returns.
Bunch:

    ---
    title: Test Shortcuts
    toggles: false
    ---
    
    (shortcut bunchtest)
    projectname = $ pbpaste
    Artikelname = ?"What's the result? (Hint: ${projectname})"

Colsole logs:

    2022-08-07 19:08:51: ⚪[shortcuts-test] Open >>>>>>>>>>>>>>>>>>>>>
    2022-08-07 19:08:55: ⚪[shortcuts-test] Processing command "shortcut bunchtest"

Shortcut won't run, pasteboard not filled. Launched from Shortcuts App works, launch from Terminal (with $TEST as output, as documented by RosemaryOrchard) works. 

If I change the Shortcuts line to

    $ /usr/bin/shortcuts run bunchtest

it works like a charm. Console logs:

    2022-08-07 19:14:29: ⚪[shortcuts-test] Open >>>>>>>>>>>>>>>>>>>>>
    2022-08-07 19:14:35: ⚪[shortcuts-test] Running Shell Script (/usr/bin/shortcuts run bunchtest)

The data gets written to the clipboard and can be processed.
I'm getting the feeling this is again one of these localization issues …

**Comment by ttscoff** [comment]
So according to the `shortcuts` man page you can use `--output-path -` to output to STDOUT, but it doesn't seem to work. I can get a result if I use an actual path, e.g. `--output-path testing.txt`, but nothing just to STDOUT, which means you and I can't pick it up in Bunch using current methods. @RosemaryOrchard have you had luck getting output on the command line?

**Comment by RosemaryOrchard** [comment]
The `Stop and Output` action is critical, it then returns to both AppleScript, and in my test just now, a variable when run through the CLI:
```sh
$  ~ TEST=$(shortcuts run "TestAppleScript")
$  ~ echo $TEST
Hi
```

![](https://user-images.githubusercontent.com/16113535/183296005-9de5ea04-51b3-4ee4-8ac8-dd19b95ef516.png)

**Reply by Younghart** [reply]
Can confirm, with the "Stop & output" action, I get a variable reliably filled on the command line. One huge step forward - and another thing learned! 

Thank you for your help (an in general – I'm an Automators.fm listener since ep01!)

**Comment by ttscoff** [comment]
Oh that's so weird that it will populate a variable but not actually output to STDOUT. (Seriously, try running `shorcuts run "TestAppleScript" -o -` and see if you get a response in Terminal). I have no idea how that could be possible. I'll have to see if I can get it to work within Bunch.

**Reply by RosemaryOrchard** [reply]
It is very weird, but it's a start at least!

**Reply by RosemaryOrchard** [reply]
Wait, you may also need to specify the [`--output-type`](https://developer.apple.com/library/archive/documentation/Miscellaneous/Reference/UTIRef/Articles/System-DeclaredUniformTypeIdentifiers.html#//apple_ref/doc/uid/TP40009259-SW1)

> If the shortcut has an action that ends with output or has a Stop and Output action, you can pipe the output of a shortcut to another command using the | character. You can define the type of output (for example, forcing the output to be a JPEG instead of text) by using the --output-type flag with a Uniform Type Indicator. To run the “Combine Images” shortcut combining all the images on the Desktop and passing to another command as an RTFD, you could use the following:
> ```sh
> shortcuts run "Combine Images" -i ~/Desktop/*.jpg --output-type com.apple.rtfd | …
> ```

From the [online help](https://support.apple.com/en-mo/guide/shortcuts-mac/apd455c82f02/5.0/mac/12.0)

**Reply by Younghart** [reply]
I would not have dreamed to rope in my two favorite automation experts/podcast hosts with one small question – on a sunday, no less! Looking forward to future developments. Thank you both so much!

**Comment by ttscoff** [comment]
Ok, I was able to get a response within Bunch, so the new syntax will allow:

```
response = (shortcut Project Setup "any input text or ${variable} you want (optional)")
```

I'll have a new beta (151) out with this addition in a little bit.

**Reply by ttscoff** [reply]
Update is out, if you're running the beta use Bunch->Check for Updates in the menu.

**Reply by Younghart** [reply]
Awesome, thank you!

**Reply by Younghart** [reply]
My first try:

    projectname = (shortcut bunchtest)
    projectadjusted = ?"What's the result? (Hint: ${projectname})"

Result:

![CleanShot 2022-08-07 at 19 35 26](https://user-images.githubusercontent.com/2934471/183303995-201b6d7b-bcf1-40e1-9428-74c594d52272.png)

Log:

    2022-08-07 19:38:23: ⚪[shortcuts-test] Open >>>>>>>>>>>>>>>>>>>>>

(Side note: I added a `(pause 10)` to no avail, deleted it again, then retried the Bunch several times. The 10 s delay would not disappear until I quit and restarted Bunch beta 151)

**Reply by Younghart** [reply]
In any event, all this is by no means urgent. I'm just over the moon that you consider a feature request by me as valid. Have a great Sunday, and most of all: Thanks 4 Bunch!

**Comment by ttscoff** [comment]
You're not on the beta. All of this is only available in beta 151.

-Brett
________________________________

**Reply by Younghart** [reply]
In between my many posts today, I did actually switch to Bunch beta, updated it and ran with it. This first try is with 1.4.10 (151). Apologies for not making that clear.

**Comment by ttscoff** [comment]
Does running the command without the variable assignment work? Just the:

(shortcut bunchtest)

part? Realizing that no other commands return values and I probably haven't set the regexes up to recognize this yet.

-Brett
________________________________

**Reply by Younghart** [reply]
Aha, different results! It works, though not in the right order. My Bunch:

    projectname = test9876
    (shortcut bunchtest)
    projectadjusted = ?"What's the result? (Hint: ${projectname})"

Result: Pop-up window for text entry (appears twice) with "Hint: test9876"
Afterwards, the shortcut is executed, and a result appears in the log (and the clipboard). Log:

    2022-08-07 21:30:20: ⚪[shortcuts-test] Open >>>>>>>>>>>>>>>>>>>>>
    2022-08-07 21:30:49: ⚪[shortcuts-test] Processing command "shortcut bunchtest"
    2022-08-07 21:30:49: ⚪[shortcuts-test] Running shortcut bunchtest
    2022-08-07 21:30:49: ⚪[shortcuts-test] Shortcut response: 2022-08-07T21:30:49+02:00

Hope this helps …

**Comment by ttscoff** [comment]
Beta 154 fixes the variable assignment with command issue.

**Reply by Younghart** [reply]
Awesome!
Here's what I did to test it with ß155:
I made a shortcut to create a folder from text it has received via quick action. It adds the folder's name to the clipboard, stops and returns the folder name:
![beta155_shortcut](https://user-images.githubusercontent.com/2934471/183411590-a64d0023-8157-401b-a8dc-9b4ed8cb32dc.png)

The accompanying Bunch looks like this:

    ---
    title: Test Shortcuts
    toggles: false
    ---
    projectdraft = "0123"
    projectname = (shortcut new Project folder "test_folder_${projectdraft}")
    pboardcontent = $ pbpaste
    
    TextEdit ~5
    - /Users/immo/Documents/Bunches/bunchtestlog.txt
    
    projectadjusted = ?"What's the result? (Hint: ${projectname} or ${pboardcontent})"

Result:
The interactive part always gets precedence, so it can't receive content from the shortcut. Actually, whatever I do, the shortcut always finishes last. This is an interesting caveat and might be worth mentioning in the Docs. But maybe this is just my Mac, another localization issue, or humidity-related.
The shortcut itself works and even receives the text sent to it, including a previously defined variable. 
One small **bug**: The folder name ends with an additional question mark – relic of a non-standard character that got somehow sent to the shortcut?
Log:

    2022-08-08 13:55:39: ⚪[shortcuts-test] Open >>>>>>>>>>>>>>>>>>>>>
    2022-08-08 13:55:43: ⚪[shortcuts-test] Running shortcut new Project folder
    2022-08-08 13:55:44: ⚪[shortcuts-test] Adding 5s delay to (TextEdit)
    2022-08-08 13:55:49: ⚪[shortcuts-test] Opening file /Users/immo/Documents/Bunches/bunchtestlog.txt with app TextEdit

tl;dr: **Getting data to and from shortcuts works in ß155**

**Comment by ttscoff** [comment]
Yeah, dialogs will always execute before commands, a consequence of the 
way I parse the Bunch. Queries are parsed first, so that optional 
snippets they cause to be included can be parsed at the same time as 
everything else. I may look into that again at some point.

When you say the "folder" has a question mark, do you mean the return 
value of the shortcut? Where are you seeing it, specifically?

**Reply by Younghart** [reply]
Ah, very well. That's something I can get behind.

The "?" appears at the end of the folder name, so it's getting passed into the shortcut – or is interpreted from the shortcut as such:
![CleanShot 2022-08-08 at 17 57 30](https://user-images.githubusercontent.com/2934471/183461095-57a20aae-0577-4e44-abad-d3da4d31df48.png)
.

**Comment by ttscoff** [comment]
Fixed, try beta 160. Also, if you're updating from 158 or later the autoupdater should be working again.

**Reply by Younghart** [reply]
Wow. Just … wow.
This opens so many opportunities! 
Bunch can feed data into shortcuts, even interactively, and get results back.

I tested a data round trip with this Bunch:

    ---
    title: Test Shortcuts
    toggles: false
    ---
    # $ /usr/bin/shortcuts run "new Project folder"
    projectdraft = ?"What's the project's name?"
    projectname = (shortcut new Project folder "test_folder_${projectdraft}")
    
    TextEdit ~5
    - /Users/immo/Documents/Bunches/bunchtestlog.txt
    
    Finder ~7
    - /Users/immo/Documents/${projectname}

The Shortcut looked like this:

![CleanShot 2022-08-08 at 22 11 13](https://user-images.githubusercontent.com/2934471/183505610-f15887e2-8152-4c16-9660-b4db30684049.png)


The whole thing works flawlessly, the folder is interactively named, the folder name makes the trip back to finally visit the newly created folder via Bunch.

Thank you, this will prove to be amazing!

**Comment by ttscoff** [comment]
Awesome, glad it's working!

-Brett

**Comment by ttscoff** [comment]
The only thing I can think of is that in order to close the STDIN pipe 
that passes the input, I have to add a newline at the end. It's possible 
this is being interpreted as a garbage character by Shortcuts, though in 
my testing I was able to pass in a string and pass it right back out, 
and what I got back was trimmed. Though, now that I say that out loud, I 
think Bunch might be doing the trimming...

I'll experiment a little further with it.

**Comment by ttscoff** [comment]
Yep, it was the newline. Found a workaround, new build will be out 
shortly.
